Mage/src/main/java/mage/watchers/common/AttackingBlockingWatcher.java

```diff
@@ -0,0 +1,68 @@
+package mage.watchers.common;
+
+import mage.constants.WatcherScope;
+import mage.game.Game;
+import mage.game.events.GameEvent;
+import mage.watchers.Watcher;
+
+import java.util.HashSet;
+import java.util.Set;
+import java.util.UUID;
+
+/**
+ * Stores attacking/blocking combat information but is only updated
+ *  1) as damage is dealt (combat damage or otherwise)
+ *  2) as a spell or ability starts resolving
+ *  3) Land playing or special action taken (just in case that then causes a static effect to kill)
+ * Thus the information is available after any involved creatures die
+ * and all dying creatures can see all other creatures that were in combat at that time
+ * @author notgreat
+ */
+public class AttackingBlockingWatcher extends Watcher {
+
+    private Set<UUID> attackers = new HashSet<>();
+    private Set<UUID> blockers = new HashSet<>();
+
+    public AttackingBlockingWatcher() {
+        super(WatcherScope.GAME);
+    }
+    @Override
+    public void watch(GameEvent event, Game game) {
+        switch (event.getType()) {
+            case LAND_PLAYED:
```

Comment:
Events are unrelated to combat at all. I know that you try to catch a nearest event before zone change on die but that smells bad with “random” events. Why not ZONE_CHANGE event to keep track of removing objects?

Anyway good implementation must use same logic and events as BlockingOrBlockedWatcher does (see original discussion above). Remove from battlefield is mtg rule’s event and must be fully supported by game engine. If it miss some use cases then must be fixed (fix original problem with remove event, not by workaround like your).

I’ll approve current code, but as a temporary solution with added a TODO task to rework that watcher later and with ZONE_CHANGE event instead lands/spells, etc.

> 506.4 A permanent is removed from combat if it leaves the battlefield, if its controller changes, if it phases out, if an effect specifically removes it from combat, if it’s a planeswalker that’s being attacked and stops being a planeswalker, or if it’s an attacking or blocking creature that regenerates (see [rule 701.13](http://blogs.magicjudges.org/rules/cr701/#13)) or stops being a creature. A creature that’s removed from combat stops being an attacking, blocking, blocked, and/or unblocked creature. A planeswalker that’s removed from combat stops being attacked.

Reply:
ZONE_CHANGE definitely wouldn't work since those are triggered sequentially alongside the trigger filter being checked, so whichever creature dies "last" would always see itself as alone. ZONE_CHANGE_GROUP might be able to be used, but with that it seems hard to make sure every attacking/blocking creature is added correctly and there might be some bad interactions with exile-on-death replacement effects and the like.

Here's the list of all watchable events that occur as three creatures die in the DoubleBlocked test (with `checking` where the filter's predicate is checked)
```
DAMAGED_BATCH_FOR_PERMANENTS
COMBAT_DAMAGE_STEP_PRIORITY
COMBAT_DAMAGE_STEP
DESTROYED_PERMANENT
DESTROYED_PERMANENT
DESTROYED_PERMANENT
ZONE_CHANGE
checking
ZONE_CHANGE
checking
ZONE_CHANGE
checking
ZONE_CHANGE_GROUP
COMBAT_DAMAGE_STEP_POST
```
During the first `DESTROYED_PERMANENT` event the game's combat info is usable, but the creatures are removed from the combat list one by one as those events are processed. By the first ZONE_CHANGE event all dying creatures are not in the combat list.

The `BlockingOrBlockedWatcher` doesn't remove any info until the entire combat is complete- which is usually fine, but using that design here would make it unable to handle the AttackMakesToken test where two different creatures are killed while "attacking alone" in the same combat.

Reply:
Watchers processing before triggers, so watcher data will be updated on ZONE_CHANGE first.

![shot_231113_085413](https://github.com/magefree/mage/assets/8344157/c435dc1f-0e1d-4f6f-8bdd-f60d7f43405d)

Reply:
The problem is the three `checking`s (inside the `triggers.checkTriggers()`) occur in three separate `handleEvent`s but all need to see the same data in the watcher about which creatures were in combat at that time.